src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,205 @@
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = node(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> n = first; n != null; n = n.next) {
+                if (n.item == null) {
+                    unlink(n);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> n = first; n != null; n = n.next) {
+                if (object.equals(n.item)) {
+                    unlink(n);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of size.");
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (!isElementPosition(index)) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of size.");
+        }
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    private boolean isElementPosition(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private void linkLast(T element) {
+        Node<T> lastNode = last;
+        Node<T> newNode = new Node<>(lastNode, element, null);
+        last = newNode;
+        if (lastNode == null) {
+            first = newNode;
+        } else {
+            lastNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkBefore(T element, Node<T> s) {
+        Node<T> pred = s.previous;
+        Node<T> newNode = new Node<>(pred, element, s);
+        s.previous = newNode;
+        if (pred == null) {
+            first = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
+    }
+
+    T unlink(Node<T> node) {
+        Node<T> prev = node.previous;
+        Node<T> next = node.next;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            node.previous = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.previous = prev;
+            node.next = null;
+        }
+        T element = node.item;
+        size--;
+        return element;
+    }
+
+    Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            Node<T> f = first;
```

Comment:
```suggestion
            Node<T> pointer = first;
```
f is not an informative name